GCS: Upgrade assistant

mlyle opened the pull request:
Provides an upgrade assistant that retrieves settings from the quad and sends them to the autotown cloud service to be adapted for the current release, and then upgrades the quad to the latest rev.

Fixes #3 
Fixes #83 
Fixes #283 

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/801)

<!-- Reviewable:end -->

mlyle committed:
gcs: rawhid: simplify signal filter

It is possible this will solve one class of problems with "multiple
devices in bootloader state," too.

mlyle committed:
gcs: uploader: add upgrade assistant stub UI classes

mlyle commented:
UI looks like this during an upgrade (will bold the operation in progress and add 'done!' to the end of a completed line).

![image](https://cloud.githubusercontent.com/assets/90903/13767036/26fed1f4-ea24-11e5-9e11-a911c2ea965f.png)

mlyle committed:
gcs: uploader/upgrader: trigger upgrade dialog

mlyle commented:
@tracernz says on IRC I should s/STM32F1/legacy/

mlyle committed:
gcs: uploader/upgrader: get dialog displayed correctly

mlyle committed:
gcs: uploader/upgrader: improve assistant step list display

mlyle commented:
This is what it looks like now:

![image](https://cloud.githubusercontent.com/assets/90903/13791869/56bb7bee-eaac-11e5-9e5a-a7871f9a2da6.png)

tracernz commented:
Niiiiiice. :D

mlyle committed:
gcs: uploader: formatting and factoring improvements

mlyle committed:
gcs: uploader/upgrader: closer to final factoring

Just need an implementation at this point...

mlyle committed:
gcs: uploader/upgrader: wire up "abort" button

mlyle committed:
gcs: uploader/upgrader: do some stuff for the first time!

mlyle committed:
gcs: uploader: improve logic for selection of fw images

1) method allows you to specify a type, like "fw", "bu", "up"--
currently unused
2) somewhat-intelligently search possible paths

mlyle committed:
gcs: uploader/upgrader: now flashes firmware

mlyle committed:
gcs: uploader/upgrader: installs firmware and boots

tracernz commented on ground/gcs/src/plugins/boards_stm/discoveryf4.cpp:
```diff
@@ -63,20 +63,6 @@ QString DiscoveryF4::boardDescription()
 //! Return which capabilities this board has
 bool DiscoveryF4::queryCapabilities(BoardCapabilities capability)
 {
```
Q_UNUSED(capability)

mlyle committed:
gcs/uploader: mostly works on CC3D

There's some ugliness about detection of devices on OSX (at least) that
needs to get figured out.

mlyle commented:
rebased

mlyle committed:
gcs: uploader/upgrader/rawhid: improve robustness

at this point I estimate I'm about 70% done with autoupgrade
functionality.

mlyle commented:
https://www.youtube.com/watch?v=jfvwO5P9KJQ

I'm about 70% done overall.  There's a bit more functional work and then a lot of test cycles remaining.

mlyle committed:
gcs: uploader/upgrader: allow user to cancel backup save

and misc cleanup

tracernz commented:
CC3D running 425feaf3

```
18:20:51[DEBUG]TelemetryMonitor: connected
18:20:51[DEBUG]"Telemetry connected"
18:20:55[DEBUG]"Sending IAP Step 1"
18:20:56[DEBUG]"Sending IAP Step 2"
18:20:56[DEBUG]"Sending IAP Step 3"
18:20:57[DEBUG]"Trying to detect bootloader"
18:20:58[DEBUG]"UpgradeAssistant: in step 0"
18:20:58[DEBUG]"UpgradeAssistant: in step 0"
18:20:58[DEBUG]"UpgradeAssistant: in step 0"
18:20:58[DEBUG]"UpgradeAssistant: in step 2"
18:20:58[DEBUG]"File loaded successfully"
18:20:58[DEBUG]""
18:20:58[DEBUG]"Starting upload..."
18:20:58[DEBUG]"Starting upload"
18:20:59[DEBUG]"Erasing, please wait..."
18:20:59[DEBUG]"[tl_dfu] Couldn't start partition upload, status: Outside Device Capabilities, additional: 0x00000700"
18:20:59[DEBUG]"Firmware upload failed"
18:20:59[DEBUG]"Upgrade assistant failed: Unable to flash upgrader image to board!"
```

tracernz commented:
Comes from here: https://github.com/d-ronin/dRonin/blob/next/flight/targets/bl/f1/op_dfu.c#L186 (additional contains the last command which is 0x07 == Upload).

mlyle commented:
Huh, very wacky.  Are you able to program firmware normally?

tracernz commented:
Yep. Upgraded bootloader to one built from your tree, same thing.

mlyle commented:
Also, OSX or Linux?

tracernz commented:
OSX

tracernz commented:
Sparky2 worked:

```
18:54:19[WARNING]"addOverrideAction: action already registered for context when registering ''"
18:54:20[DEBUG]Automatically opening device
18:54:20[WARNING]QSslSocket: cannot call unresolved function SSL_get0_next_proto_negotiated
18:54:20[WARNING]QSslSocket: cannot call unresolved function SSL_get0_next_proto_negotiated
18:54:20[WARNING]QSslSocket: cannot call unresolved function SSL_get0_next_proto_negotiated
18:54:21[DEBUG]ConnectionManager::devChanged autoconnected USB device
18:54:21[DEBUG][SystemHealth] Warning: The SystemHealth SVG does not contain a graphical element for the  "Geofence"  alarm.
18:54:21[DEBUG][SystemHealth] Warning: The SystemHealth SVG does not contain a graphical element for the  "GyroBias"  alarm.
18:54:21[DEBUG]Connection with the autopilot established
18:54:23[DEBUG]TelemetryMonitor: connected
18:54:23[DEBUG]"Telemetry connected"
18:54:25[DEBUG]"Sending IAP Step 1"
18:54:26[DEBUG]"Sending IAP Step 2"
18:54:26[DEBUG]"Sending IAP Step 3"
18:54:27[DEBUG]"Trying to detect bootloader"
18:54:28[DEBUG]"UpgradeAssistant: in step 0"
18:54:28[DEBUG]"UpgradeAssistant: in step 0"
18:54:28[DEBUG]"UpgradeAssistant: in step 0"
18:54:28[DEBUG]"UpgradeAssistant: in step 4"
18:54:28[DEBUG]""
18:54:28[DEBUG]"Downloading settings partition..."
18:54:56[DEBUG]"UpgradeAssistant: in step 5"
18:55:02[DEBUG]"Retrieved dump of configuration from cloud"
18:55:09[DEBUG]"Dump of configuration saved to file!"
18:55:09[DEBUG]"UpgradeAssistant: in step 6"
18:55:09[DEBUG]"UpgradeAssistant: in step 8"
18:55:09[DEBUG]"File loaded successfully"
18:55:09[DEBUG]""
18:55:09[DEBUG]"Starting upload..."
18:55:12[DEBUG]"Starting upload"
18:55:13[DEBUG]"Erasing, please wait..."
18:55:15[DEBUG]"Uploading firmware partition..."
18:55:38[DEBUG]"Firmware upload success"
18:55:38[DEBUG]"Starting firmware metadata upload"
18:55:38[DEBUG]"Starting upload"
18:55:38[DEBUG]"Erasing, please wait..."
18:55:38[DEBUG]"Uploading description partition..."
18:55:38[DEBUG]"Firmware and firmware metadata upload success"
18:55:38[DEBUG]"UpgradeAssistant: in step 9"
18:55:39[DEBUG]"Bootloader disconnection detected"
18:55:40[DEBUG]Automatically opening device
18:55:40[DEBUG]ConnectionManager::devChanged autoconnected USB device
18:55:40[DEBUG]Connection with the autopilot established
18:55:41[DEBUG]TelemetryMonitor: connected
18:55:42[DEBUG]"Telemetry connected"
18:55:42[DEBUG]"File loaded successfully"
18:55:42[DEBUG]"UpgradeAssistant: in step 10"
```

mlyle committed:
gcs: uploader/upgrader: detect lack of settings part

mlyle committed:
gcs: uploader/upgrader: open all firmware files early

To detect errors before we do destructive operations.

mlyle committed:
gcs: uavsettingsimportexportfactory->uavsettingsimportexportmanager

mlyle committed:
gcs: uploader/upgrader: hook import functionality in

mlyle committed:
gcs: uploader/upgrader: set ignored githash properly

mlyle committed:
gcs: cleanup to upgrade/import functionality

mlyle committed:
gcs: uavsettingsimportexport: fix line ends

mlyle committed:
gcs: uploader/upgrader: signal completion

mlyle committed:
flight: bu: do things much faster

Some of these delays were incredibly reckless and increased exposure to
the things that can brick a board.  The worst?  A delay between erasing
flash and programming it.

mlyle committed:
gcs: uploader/upgrader: properly catch loader

mlyle committed:
gcs: uploader/upgrader: fix depends

mlyle committed:
gcs: uavsettingsimport: do error box when no config in dump

mlyle committed:
gcs: uploader/upgrader: resolve a few remaining issues

Get the real revisions; add firmware flashing timeout

mlyle committed:
gcs: uploader/upgrader: minor code reorg

mlyle commented:
Putting this up for review.  I assume we'll have a couple of PRs after this one on this track.

We still need to identify @tracernz 's problem on CC3D.

mlyle committed:
gcs: uploader: avoid duplicate operation on connect

mlyle commented:
jenkins, build artifacts please

mlyle committed:
gcs: usbmonitor: fix windows build

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
Tried on windows, seems to work on simple test cases.  The upgrader status window sometimes ends up under all the other windows though which sucks.

mlyle committed:
gcs: uploader/upgrader: file extension xml->uav

mlyle committed:
gcs: uploader/upgrader: increase timeouts

Feedback from @yamahabest that he has a Windows laptop that hangs for 18
seconds on board connection; wait 25 seconds for boot steps as a result.

mlyle committed:
gcs: upgrade: keep window on top at step change

Seems to mostly affect Windows.  Effort to make things more reasonable.

mlyle commented:
jenkins, build artifacts please

pug398 commented:
"Unable to load legacy upgrader image for board"

Is this what @tracernz was getting on CC3D?

mlyle commented:
No, tracernz was getting timeouts. That sounds like a packaging issue-- maybe i'm not packaging the up_cc3d.tlfw image.  One second.

mlyle commented:
P.S. I mostly want this to work well from endomorphism or Renatus-- I"m not sure how it will work from random next builds at the moment.

mlyle commented:
Argh.

@pug398-- are you using the windows zip image?  It seems we don't know how to load the firmware from that zip package yet.

if so, can you try moving firmware_20160318-56b8d105 into gcs_winx86_amd64-20160318-56b8d105 and renaming the folder to be named "firmware" ?

pug398 commented:
OK that let it see the firmware and it started the updating process. It got to booting...importing settings...
I could see metadata loaded successfully in background but then it hung and the same message above appeared again. Rescue, boot and firmware was on the board.

mlyle committed:
package: alter windows packaging of firmware

To put it in a path where we can find it.

mlyle commented:
@pug398 -- can you get debug output from a failed run?

Also be careful-- that the import dialog isn't trying to activate but in the wrong place / under other windows.. I'm fighting the window stacking behavior.

mlyle commented:
jenkins, build artifacts please

pug398 commented:
Just got back into the board after several rescue attempts. It was stopping at downloading partition...
Got in and flashed the firmware manually and now it will connect. Will make another attempt here shortly.

mlyle commented:
http://pastebin.com/hJrkEzuQ

mlyle committed:
gcs: uploader: fix export filename

pug398 commented:
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta name="qrichtext" content="1" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body style=" font-family:'MS Shell Dlg 2'; font-size:8.25pt; font-weight:400; font-style:normal;">

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:23:55[DEBUG]&quot;No devices in bootloader state detected&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:23:55[DEBUG]&quot;GCS ready for flight board connection...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:24:54[DEBUG]&quot;Downloading description partition...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;Connection to bootloader successful&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;Ready to flash firmware&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;UpgradeAssistant: in step 0&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;UpgradeAssistant: in step 0&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]Upgrading from  &quot;ea015996&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;UpgradeAssistant: in step 0&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;UpgradeAssistant: in step 2&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;File loaded successfully&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;Starting upload...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:01[DEBUG]&quot;Starting upload&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:02[DEBUG]&quot;Erasing, please wait...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:04[DEBUG]&quot;Uploading firmware partition...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:05[DEBUG]&quot;Firmware upload success&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:05[DEBUG]&quot;Starting firmware metadata upload&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:05[DEBUG]&quot;Starting upload&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:06[DEBUG]&quot;Erasing, please wait...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:06[DEBUG]&quot;Uploading description partition...&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:06[DEBUG]&quot;Firmware and firmware metadata upload success&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:06[DEBUG]&quot;UpgradeAssistant: in step 3&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:25:31[DEBUG]&quot;Upgrade assistant failed: Unable to enter legacy upgrade tool on board!!&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:26:01[DEBUG]&quot;UpgradeAssistant: in step 4&quot;</span></p>

<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#0000ff;">19:26:01[DEBUG]&quot;Cannot perform the selected operation if not in bootloader state&quot;</span></p></body></html>

mlyle commented:
eek.  What Windows version?  CC3D, I assume?

pug398 commented:
win7 and CC3D. Since I don't need a map I could try win10 I guess :)

pug398 commented:
and the file created is suppose to be cloud_exported.uav.xml? I always get a little nervous when I see 2 dots in a file name

mlyle commented:
the .uav.xml is a bug.  it should be fixed in the most recent one.

We don't use the file though-- it's just a backup of the configuration if the rest of the upgrade goes wrong.

pug398 commented:
Yeah it looks fixed now. quanton loaded and booted and so did sparky1. They didn't release at the end and required a power cycle to reconnect but they did upgrade. Revolution fails to upload firmware partition.

mlyle committed:
ground: upgrader: stagger 400ms delay between most steps

Makes it feel better.  Does it make things more reliable for
@tracernz?

mlyle committed:
gcs: tl_dfu: put back a possibly-functional status req

mlyle committed:
gcs: usbmonitor: replace with dirt simple polling impl

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle committed:
usbmonitor: remove usage/usagepage fields, add debug

mlyle committed:
usbmonitor: hack for Linux enumeration peculiarities

mlyle commented:
jenkins, build artifacts please

mlyle committed:
gcs: tl_dfu/uploader: debugging

mlyle commented:
jenkins, build artifacts please

mlyle commented:
Just a note to everyone-- this is no good at the moment.  I'm building artifacts to hopefully investigate a little tomorrow.

mluessi commented:
Very nice. Getting rid of the separate usb implementations for Win/Linux/OSX is great. One suggestion for the UI: how about hiding the unnecessary steps for non F1 targets? I think it would make it looks cleaner / simpler and hopefully the majority of users will be on non-F1 boards in the near future.

mlyle commented:
> . One suggestion for the UI: how about hiding the unnecessary steps for non F1 targets? I think it would make it looks cleaner / simpler and hopefully the majority of users will be on non-F1 boards in the near future.

I originally had it that way and @tracernz asked me to do the opposite.  You guys decide what you want ;)

mlyle committed:
gcs: hidapi enumeration changes that may improve win/linux

mlyle commented:
Woudln't be surprised if I have a compile error in this, or if it doesn't help.  Yay.

jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

(that was dumb)

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
sorry for spam everyone, making the same change 3 times sucks, and i'm not doing very well at it :P 

jenkins, build artifacts please

mluessi commented:
@tracernz why do you think always showing the steps that will be skipped on non F1 platforms is better?

Something else: I think we should have an option to not port settings and instead initialize all settings to the new defaults. The rationale is as follows: When making new boards, they are usually loaded with the dRonin version that is stable at the time of testing. The units then get shipped to a warehouse/distributor and may sit there for a while before the end users receive them. So, when users first connect their flight controller to the GCS, it will have to be updated to the latest release. In this case, the settings shouldn't be ported and the new defaults should be used. Maybe the assistant could ask, "Do you want to port existing settings? Do not use this option when setting up a new flight controller that has not previously been configured" YES/NO. Thoughts?

mlyle commented:
@mluessi -- if you ship an empty settings partition I'll handle it soon (this does dumb things right now when the flight controller starts off erased).

mluessi commented:
@mlyle but the settings partition will be filled when the FC first boots, no? So unless a user uses rescue mode and immediately connects it to the updater, it won't work.

mluessi commented:
Or maybe I'm wrong, i.e., settings are only written to flash when they are saved from the GCS. Is this the case?

mlyle commented:
That's the case, with the exception of things like systemident.. but you'd have to fly for that to happen.

mlyle commented:
i kan prorgram gud

jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

tracernz commented:
@mluessi I think it's best to have a consistent UI whose elements don't move around during runtime. I think hiding things only confuses people when they sometimes see them and sometimes don't, whereas an explicit clue like (skipped) makes it clearer.

tracernz commented:
@mlyle Don't seem to get the events that trigger uploader to connect to the board anymore on Linux? USBMonitor sees the devices being added and removed but nothing ever happens until it boots into firmware, then it connects as usual.

mlyle commented:
jenkins, build artifacts please

mlyle committed:
upgrade: some changes for linux

Cope better with timeouts sending frames; we're doing stuff more
aggressively now at times so that can happen.

Add a delay at reset from bootloader (not jumptoapp) to ensure that
we re-enumerate properly.

Return errorcodes from linux hidapi.

mlyle committed:
gcs: uploader/tl_dfu: add logic to resume download

Necessary on slow machines to be 100% reliable.

Else it usually works, but occasionally doesn't.

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle committed:
gcs: uploader: small changes to timeouts and debug msgs

mlyle commented:
jenkins, build artifacts please

tracernz commented:
This works for me on Linux/CC3D now.

mlyle commented:
@pug398 if you can easily try this now I'd really appreciate it-- this has been a difficult one to get mostly correct.  I have a set of known issues / things I anticipate doing in follow-up PRs in #821.

tracernz commented on ground/gcs/src/plugins/uploader/tl_dfu.cpp:
```diff
@@ -248,12 +256,29 @@ bool DFUObject::DownloadPartition(QByteArray *fw, qint32 const & numberOfBytes,
+                    AbortOperation();
+
+                    // Drain out pending packets.
+                    for (unsigned int i=0; i<(msg.numberOfPackets + 10); i++) {
```
Some 0x20s missing.

tracernz commented on ground/gcs/src/plugins/uploader/tl_dfu.h:
```diff
@@ -135,7 +135,7 @@ public slots:
     // USB coms:
     int SendData(bl_messages);
-    int ReceiveData(bl_messages &data);
+    int ReceiveData(bl_messages &data, int timeoutMS=10000);
```
whitespace is your friend :P

mlyle replied:
no, no it's really not, but fixing :P

mlyle committed:
gcs: uploader: whitespace fixen from review

mlyle committed:
gcs: uploader: ensure firmware is properly "loaded" at conn-time

In non-upgrade mode.

pug398 commented:
Can do but it will be about 5 hours before I can get to it. You want windows test of just CC3D or broader test?

mlyle commented:
@pug398 whatever you can do, I really appreciate (though if it sucks right away, please give up :)

mlyle committed:
gcs: usb: small improvements

* filter enumeration for telemetry to devices not in bootloader state
* improve debug output
* prevent duplicate enumeration on Windows

mlyle commented:
jenkins, build artifacts please

tracernz commented on ground/gcs/src/plugins/rawhid/rawhidplugin.h:
```diff
@@ -95,7 +92,7 @@ class RAWHID_EXPORT RawHIDPlugin
     virtual void extensionsInitialized();
 private:
-    RawHIDConnection *hidConnection;
+    RawHIDConnection* hidConnection;
```
Y you do dis?

mlyle replied:
It's supposedly qt style to group with the type name and is consistent with
the other declarations nearby I think.

I hate it too.

On Mon, Mar 21, 2016, 12:04 PM Michael Corcoran notifications@github.com
wrote:

> In ground/gcs/src/plugins/rawhid/rawhidplugin.h
> https://github.com/d-ronin/dRonin/pull/801#discussion_r56878221:
> 
> > @@ -95,7 +92,7 @@ class RAWHID_EXPORT RawHIDPlugin
> >      virtual bool initialize(const QStringList &arguments, QString *error_message);
> >      virtual void extensionsInitialized();
> >  private:
> > -    RawHIDConnection *hidConnection;
> > -    RawHIDConnection\* hidConnection;
> 
> Y you do dis?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/d-ronin/dRonin/pull/801/files/e0dc6bbbe03abea8536522066c3a379988660d2a..a179530519cea192514f036b123a7a25e370b8ee#r56878221

tracernz replied:
Qt coding style = \* with the symbol according to https://wiki.qt.io/Qt_Coding_Style

mlyle replied:
Ah.  I think someone @ Tau asked me to do it the other way.  That's a relief.

I will make this PR mostly conform to the right way and follow the right way goin' forward.

mlyle committed:
gcs: style: adjust grouping of * / & (per review fdbk)

pug398 commented:
This is going to be pretty neat. I had a CC3D, Sparky1 and Quanton handy so they are the first group to test. The board detection, booting into it, writing firmware all seem to go smoothly. A .uav file was written for each one. The problem for all 3 was none of them booted the final time to import settings and timed out. A quick disconnect/reconnect and they came right up to do a manual import. They all connected quickly to start the upgrade process so this has the potential to be the best it has been yet. Just need to get that final boot back into the firmware. I will check sparky2, revo and brain after you have a chance to think on it.

Ok revo was smooth including the final boot back in but failed import with "failed XML dump" message.

mlyle commented:
@pug398-- thanks for the test..  Concerned that it's not reliable on the final detection for you.  Can you get debug output for one with it failing there?

Did the revo have settings?  If it has a blank settings partition we have a bug with that now.

pug398 commented:
Yeah the revo was blank so that may be it. Will throw some settings on there and try again. It really is getting the firmware on all of them quickly and not a glitch in any of the 4 as far as that goes.

tracernz commented:
Should probably check if the XML is empty and skip the import step, and give user appropriate warning at the end? Could come in another PR.

mlyle commented:
@tracernz -- yes, it's listed in #821

pug398 commented:
Blank settings was the issue with revo. It has booted and imported several times without any problem. The previous three are still stopping at final Booting... 
Here is output of sparky1

20:37:02[DEBUG]"No devices in bootloader state detected"
20:37:02[DEBUG]"GCS ready for flight board connection..."
20:37:18[DEBUG]Adding  8352 16848 1 "36002C000257325237313620+BL" "Sparky" "Tau Labs"
20:37:18[DEBUG]usbmonitor detection cycle complete.
20:37:18[DEBUG]"EnterDFU:64 bytes sent"
20:37:18[DEBUG]"StatusRequest:64 bytes sent"
20:37:18[DEBUG]"StatusRequest:64 bytes received"
20:37:18[DEBUG]"Status:0"
20:37:18[DEBUG]FINDDEVICES BEGIN
20:37:18[DEBUG]"FINDDEVICES SENDING CAPABILITIES REQUEST BUFFER_SIZE:64"
20:37:18[DEBUG]"FINDDEVICES CAPABILITIES REQUEST BYTES_SENT:64"
20:37:18[DEBUG]"FINDDEVICES CAPABILITIES ANSWER BYTES_RECEIVED:64"
20:37:18[DEBUG]"Device ID=34818"
20:37:18[DEBUG]"Device SizeOfCode=212892"
20:37:18[DEBUG]"Device SizeOfDesc=100"
20:37:18[DEBUG]"BL Version=131"
20:37:18[DEBUG]"FW CRC=2453654102"
20:37:18[DEBUG]"Partition 0 Size 212892"
20:37:18[DEBUG]"Partition 1 Size 100"
20:37:18[DEBUG]"Partition 2 Size 16384"
20:37:18[DEBUG]"Partition 3 Size 16384"
20:37:18[DEBUG]"Partition 4 Size 16384"
20:37:18[DEBUG]"Partition 5 Size 0"
20:37:18[DEBUG]"Partition 6 Size 0"
20:37:18[DEBUG]"Partition 7 Size 0"
20:37:18[DEBUG]"Partition 8 Size 0"
20:37:18[DEBUG]"Partition 9 Size 0"
20:37:18[DEBUG]"EnterDFU:64 bytes sent"
20:37:18[DEBUG]"Downloading description partition..."
20:37:18[DEBUG]"StartDownload 2 Last Packet Size 11 Bytes sent 64"
20:37:18[DEBUG]"StatusRequest:64 bytes sent"
20:37:18[DEBUG]"StatusRequest:64 bytes received"
20:37:18[DEBUG]"Status:0"
20:37:18[DEBUG]"Download operation completed-- 100 bytes"
20:37:37[DEBUG]"Connection to bootloader successful"
20:37:37[DEBUG]"File loaded successfully"
20:37:37[DEBUG]Padding
20:37:37[DEBUG]"Ready to flash firmware"
20:37:37[DEBUG]"UpgradeAssistant: in step 0"
20:37:37[DEBUG]"UpgradeAssistant: in step 0"
20:37:37[DEBUG]Upgrading from  "7fe151ab"
20:37:37[DEBUG]"UpgradeAssistant: in step 0"
20:37:37[DEBUG]"UpgradeAssistant: in step 4"
20:37:37[DEBUG]""
20:37:37[DEBUG]"DOWNLOAD PARTITION- SETTING DOWNLOAD CONFIG PARTITION=3 SIZE:16384"
20:37:37[DEBUG]DOWNLOAD THREAD STARTED
20:37:37[DEBUG]"EnterDFU:64 bytes sent"
20:37:37[DEBUG]"Downloading settings partition..."
20:37:37[DEBUG]"StartDownload 293 Last Packet Size 8 Bytes sent 64"
20:37:38[DEBUG]"StatusRequest:64 bytes sent"
20:37:38[DEBUG]"StatusRequest:64 bytes received"
20:37:38[DEBUG]"Status:0"
20:37:38[DEBUG]"Download operation completed-- 16384 bytes"
20:37:38[DEBUG]DOWNLOAD FINISHED
20:37:38[DEBUG]downloadFinished in dSettings:  true
20:37:38[DEBUG]"UpgradeAssistant: in step 5"
20:37:41[DEBUG]"Retrieved dump of configuration from cloud"
20:37:44[DEBUG]"Dump of configuration saved to file!"
20:37:44[DEBUG]"UpgradeAssistant: in step 6"
20:37:44[DEBUG]"UpgradeAssistant: in step 8"
20:37:44[DEBUG]"File loaded successfully"
20:37:44[DEBUG]Padding
20:37:44[DEBUG]""
20:37:44[DEBUG]"Starting upload..."
20:37:44[DEBUG]"StatusRequest:64 bytes sent"
20:37:44[DEBUG]"StatusRequest:64 bytes received"
20:37:44[DEBUG]"Status:0"
20:37:44[DEBUG]Starting Firmware Upload...
20:37:44[DEBUG]"Bytes Loaded=197628"
20:37:44[DEBUG]Padding
20:37:44[DEBUG]"NEW FIRMWARE CRC=2917170212"
20:37:44[DEBUG]"Number of packets:3530 Size of last packet:1"
20:37:44[DEBUG]"Starting upload"
20:37:45[DEBUG]"64 bytes sent"
20:37:45[DEBUG]Erasing memory
20:37:45[DEBUG]"Erasing, please wait..."
20:37:47[DEBUG]"StatusRequest:64 bytes sent"
20:37:47[DEBUG]"StatusRequest:64 bytes received"
20:37:47[DEBUG]"Status:1"
20:37:47[DEBUG]"StatusRequest:64 bytes sent"
20:37:47[DEBUG]"StatusRequest:64 bytes received"
20:37:47[DEBUG]"Status:1"
20:37:47[DEBUG]"Erase returned:Uploading"
20:37:47[DEBUG]"Start Uploading:3530 56 byte packets"
20:37:47[DEBUG]"Uploading firmware partition..."
20:38:01[DEBUG]"64 bytes sent"
20:38:01[DEBUG]"StatusRequest:64 bytes sent"
20:38:01[DEBUG]"StatusRequest:64 bytes received"
20:38:01[DEBUG]"Status:5"
20:38:01[DEBUG]"Status=Last operation success"
20:38:01[DEBUG]Firmware Uploading succeeded
20:38:01[DEBUG]UPLOAD FINISHED
20:38:01[DEBUG]"Firmware upload success"
20:38:01[DEBUG]"Starting firmware metadata upload"
20:38:01[DEBUG]"StatusRequest:64 bytes sent"
20:38:01[DEBUG]"StatusRequest:64 bytes received"
20:38:01[DEBUG]"Status:5"
20:38:01[DEBUG]Starting Firmware Upload...
20:38:01[DEBUG]"Bytes Loaded=100"
20:38:01[DEBUG]"NEW FIRMWARE CRC=2423730980"
20:38:01[DEBUG]"Number of packets:2 Size of last packet:11"
20:38:01[DEBUG]"Starting upload"
20:38:01[DEBUG]"64 bytes sent"
20:38:01[DEBUG]Erasing memory
20:38:01[DEBUG]"Erasing, please wait..."
20:38:01[DEBUG]"StatusRequest:64 bytes sent"
20:38:01[DEBUG]"StatusRequest:64 bytes received"
20:38:01[DEBUG]"Status:1"
20:38:02[DEBUG]"StatusRequest:64 bytes sent"
20:38:02[DEBUG]"StatusRequest:64 bytes received"
20:38:02[DEBUG]"Status:1"
20:38:02[DEBUG]"Erase returned:Uploading"
20:38:02[DEBUG]"Start Uploading:2 56 byte packets"
20:38:02[DEBUG]"64 bytes sent"
20:38:02[DEBUG]"StatusRequest:64 bytes sent"
20:38:02[DEBUG]"StatusRequest:64 bytes received"
20:38:02[DEBUG]"Status:5"
20:38:02[DEBUG]"Status=Last operation success"
20:38:02[DEBUG]Firmware Uploading succeeded
20:38:02[DEBUG]UPLOAD FINISHED
20:38:02[DEBUG]"Uploading description partition..."
20:38:02[DEBUG]"Firmware and firmware metadata upload success"
20:38:02[DEBUG]Padding
20:38:02[DEBUG]"05854b27"
20:38:02[DEBUG]"UpgradeAssistant: in step 9"
20:38:27[DEBUG]"Upgrade assistant failed: Unable to connect to new firmware!"

pug398 commented:
I think the issue is with little quirks within each firmware that have always been causing a problem and this is just making them rear their ugly head. Sparky1 can be made to update and boot by putting on an old firmware and erasing settings partition. Import settings and then immediately switch to the new updater and it completes. If you switch back to old firmware and come back to new updater with settings already applied it will fail. Quanton is throwing an error code (4 flash) when the settings are erased during updating. When it throws the error you immediately reset board and everything will complete correctly.

mlyle commented:
@pug398-- let's just stick with one case, because I think a lot of this is variations on the same theme, just manifesting differently.   I have one more dumb question with the sparky1 case above.  Do you happen to know offhand if during those last 25 seconds where it's supposed to be rebooting-- whether it's doing the bootloader fadey-thing or the firmware disarmed flashing?  Curious whether it is failing to reboot, or whether the GCS is failing to notice it's booted into the firmware.

tracernz commented:
Works for me perfectly on Linux with Sparky, including all settings imported properly.

tracernz commented on flight/targets/bl/common/main.c:
```diff
@@ -449,9 +459,10 @@ int main(void)
 		/* User has requested that we boot into firmware */
 		PIOS_IAP_ClearRequest();
-		PIOS_DELAY_WaitmS(1000);//needed so OS can detect BL USB disconnect
+		PIOS_DELAY_WaitmS(1000);//needed so OS can detect FW USB disconnect
```
Is this comment change correct?

mlyle replied:
Yes, I think so.  We're immediately going around, so this avoids us missing the vid/pid coming back with the same usage.

tracernz commented on flight/targets/bl/common/main.c:
```diff
@@ -596,6 +607,9 @@ static void process_packet_rx(struct bl_fsm_context * context, const struct bl_m
 		bl_fsm_inject_event(context, BL_EVENT_JUMP_TO_APP);
 		break;
 	case BL_MSG_RESET:
+		PIOS_IRQ_Disable();
```
Is this to force USB to re-enumerate?

mlyle replied:
Yes.

Re this and the previous comment-- note that this stuff preceded some of my other changes to the enumeration machinery, and may not be strictly necessary.  But it improved behavior with the old stuff and probably helps in some cases still.

tracernz commented on ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexport.pluginspec:
```diff
@@ -1,12 +1,12 @@
+        <dependency name="UAVObjects" version="1.0.0"/>
+        <dependency name="UAVObjectUtil" version="1.0.0"/>
+    </dependencyList>
+</plugin>
```
Line endings changed? If they're changed to Unix, good. :+1:

mlyle replied:
Yes :)

tracernz commented on ground/gcs/src/plugins/uploader/tl_dfu.cpp:
```diff
@@ -698,7 +732,21 @@ int DFUObject::SendData(bl_messages data)
+
+    int ret;
+
+    for (int i=0; i<10; i++) {
```
Missing spaces

tracernz commented on ground/gcs/src/plugins/uploader/tl_dfu.cpp:
```diff
@@ -549,7 +581,9 @@ tl_dfu::Status DFUObject::UploadPartition(QByteArray &sourceArray, dfu_partition
     }
-    ret = StatusRequest();
+
+    ret=StatusRequest();
```
Missing spaces

tracernz commented on ground/gcs/src/plugins/uploader/tl_dfu.cpp:
```diff
@@ -427,27 +463,23 @@ bool DFUObject::OpenBootloaderComs(USBPortInfo port)
     m_hidHandle = hid_open(port.vendorID, port.productID, NULL);
-    if ( m_hidHandle )
-    {
+    if ( m_hidHandle ) {
```
Remove spaces inside ()

tracernz commented on ground/gcs/src/plugins/uploader/uploadergadgetwidget.cpp:
```diff
@@ -687,35 +682,498 @@ void UploaderGadgetWidget::onExportButtonClick()
+
+void UploaderGadgetWidget::upgradeError(QString why)
+{
+    (void) why;
```
Q_UNUSED

mlyle committed:
gcs: uploader: spacing/style changes from fdbk

mlyle commented:
forcepush-- minor conflict

mlyle commented:
jenkins, build artifacts please

pug398 commented:
The bug in quanton is going to have to be fixed before this will work on it. It will not connect to a board in error state which occurs on quanton if you erase settings partition. I can make it work on sparky everytime too if I follow a certain procedure. @tracernz if you downgrade and then upgrade sparky without manually erasing settings will it still upgrade the second time?

mlyle commented:
@pug398-- did you see my question above about the sparky1 failure?  

> I have one more dumb question with the sparky1 case above. Do you happen to know offhand if during those last 25 seconds where it's supposed to be rebooting-- whether it's doing the bootloader fadey-thing or the firmware disarmed flashing?

pug398 commented:
Cloud services are blocked by firewall at current location so will have to check later.

pug398 commented:
Tried a second sparky1 that has been up and flying on later next just to be sure. It appears updater releases and board goes into normal running led flashing 1 time per second. Did not detect and start import.

mlyle commented:
@pug398 thanks--- I will think about what I can do to make the detection of the bootloader leaving more reliable.  (or a fallback to ask the user to unplug and plug in the board if there's a failure to detect at this step).

I also can't repro the Quanton thing-- just went through 2 upgrade cycles.  4 blinks is a failure to init the barometer.  Does Quanton boot for you subsequently-- it's just the boot after resetting settings that's glitchy?

mlyle committed:
gcs: upgrader: add fallback for common failure case

mlyle commented:
jenkins, build this please

Implemented (but was not able to really test) a dialog that pops up if we are unable to detect board after reset... asking user to unplug and plug in USB.

mlyle commented:
ugh, jenkins, build artifacts please is what i mean

joeynuggetz commented:
Other than a the following screenshot showing the usb connect dialog, CC3D upgraded fine.

[debuglog.txt](https://github.com/d-ronin/dRonin/files/185367/debuglog.txt)
![image](https://cloud.githubusercontent.com/assets/14797846/13975334/ef7faac8-f081-11e5-86cb-25a5d7d46059.png)

![image](https://cloud.githubusercontent.com/assets/14797846/13975339/fbf93260-f081-11e5-9f44-3d6e77b00013.png)

pug398 commented:
@mlyle The quanton board I am testing will flash baro error (4x) every time settings are erased. Does it in next with setting partition erase so that is why I said it didn't appear to be the upgrader. Reset it and all is good with no baro error indicated. Only get the error when settings are erased and that is the point it occurs in the updater steps. I could not boot back into the board on next either without a reset. May be the board but I didn't think so.

mlyle commented:
@pug398 --- Please clarify--- my Quanton doesn't panic in these scenarios and I can't see an obvious reason it would in the schematic, except maybe a weak 3V3 regulator--
1. If you select to erase the partition, does the operation succeed?  Does it remain in the loader here OK?
2. When you select to boot, is this when it flashes 4 times in a panic loop?
3. When you subsequently reset, does it still flash 4 times in a panic loop?

If it panics on step 2 and not 3, I think the change I just made may help (it instructs you to unplug and plug the board to get the reset instead, and sets things up so it won't catch the bootloader at this step).

mlyle commented:
User yamahabest from IRC reports success with Windows + Brain, other than the complaint about an empty board at the end (expected).

mlyle committed:
package: linux: adjust firmware / packaging path

So that upgrade can find the images.

mlyle commented:
jenkins, build artifacts please

jpahka commented:
Tested the latest artifacts on sparky2 on taulabs endo release, linux host.

First try didn't find the firmware images. After moving them to the right location on disk, worked perfect apart from not restoring settings for some modules, which I belive is a known problem.

Good stuff

mlyle commented:
Linux now appears to put the firmware artifacts in the right place

pug398 commented:
@mlyle 
  For Quanton:
1.  If you select to erase the partition, does the operation succeed? Does it remain in the loader here OK?
2.  When you select to boot, is this when it flashes 4 times in a panic loop?
3.  When you subsequently reset, does it still flash 4 times in a panic loop?

1 yes
2 yes
3 no

Taulabs and Renatus will boot whether you erase settings or not. Both ways there is a noticeable delay in led and board boot. New dronin flash and boot works and there is slightly less delay in board boot but it does boot. New dronin flash, erase settings,boot there appears to be no delay in board trying to boot. Leds are on immediately. Perhaps it is beating baro initialization. Do you use a settings erased flag to skip some boot code?

mlyle commented:
>  Do you use a settings erased flag to skip some boot code?

No. 

But it's possible booting takes a different amount of time whether there's settings or not-- it should be slower if it's blank, actually, because the flash filesystem will write out a first arena.  

There's also a panic code of 6 blinks if we fail to mount the filesystem.

Does the new code work?  If the target board is stuck blinking or hasn't enumerated, it should tell you to power-cycle the board-- though no one has hit this message yet (and I'm not absolutely positive that I did the right thing for when the board is completely power cycled).

mlyle committed:
pios_hal: panic: flash code 3x; reset and try again

This is an effort to mitigate times when the controller panics starting
up for whatever reason.  I've seen sparky2 to end up with
MPU-9250-related panics after many consecutive reboots; @pug398 has a
Quanton that likes to panic after settings are erased.

mlyle commented:
jenkins, build artifacts please

mlyle commented:
New refinement:  Only flash panic code 3x, then automatically reset.  So occasional glitches during startup don't leave the board panicked forever.  I've seen Sparky2 to sporadically panic on failure to init MPU-9250 on reset, and there's this init-ms5611 failure on @pug398 's quanton.

pug398 commented:
The last one appeared to work as you intended and will give this one a go when done. Sounds like it should do the trick. Hopefully it will fix up my 2 quanton boards with the strange behavior.

pug398 commented:
That worked for the quanton. Flashed and imported settings fairly quick too with brief wait for the panic code.

mlyle commented:
Nice!  Thank you for giving it a try.

mlyle commented:
@pug398 -- did the previous fix allow sparky1 to complete, too?

pug398 commented:
yeah I have done CC3D, sparky1, sparky2, quanton, revo and brain and they imported settings. The CC3D had to make use of the "click ok and power cycle" message but it imported everything ok. Sometimes a "can't connect to:" message is left up and you must close that before importing but in the end all the settings were there.

mlyle commented:
Current status of what people have tried in test (unchecked are things that I "selected" for test runs from user prevalence but no one has run yet...)

Not counting any of my own testing.
- [x] Lin+Brain mluessi - succcess
- [x] Lin+CC3D - tracernz - success
- [x] Lin+CC3D - mluessi - success
- [ ] Lin+Sparky
- [ ] Mac+Brain
- [ ] Mac+CC3D
- [ ] Mac+CC3D
- [ ] Mac+Sparky2
- [x] Win+Brain - ufoDziner - success/missing module config (3 boards/3 attempts)
- [x] Win+Brain - yamahabest - success
- [x] Win+Brain - pug398 - success
- [x] Win+Brain - droidicus - success
- [x] Win+CC3D - nuggetz - success (annoying dialog popup).
- [x] Win+CC3D - pug398 - success (required manual power cycle at end)
- [x] Win+CC3D - ufodziner - success (2 boards/2 attempts)
- [x] Win+Revo - pug398 - success
- [x] Win+Sparky - pug398 - success
- [x] Win+Sparky2 - pug398 - success
- [x] Win+Sparky2 - ufodziner - success
- [x] Win+Sparky2 - droidicus - success
- [x] Win+Quanton - pug398 - success (Panics during the restart process, but with new panic logic recovers)

Win+AQ32 - yamahabest - failure (gets stuck for 15s+ on enum on his laptop)

mlyle commented:
Thanks to everyone --- @pug398 especially -- for test and helping me work through this.

mluessi commented:
Successfully tested upgrading Brain from Renatus on Linux. Nice work, this is a huge usability  improvement.

I also tested upgrading from an earlier Brain release (Brain-20150928-R1) and some settings, e.g., ManualControlSettings, are missing. I assume it's because the cloud service doesn't know the correct git hash. Is it possible to add them? The majority of brain users still hasn't moved to dRonin and hopefully they will do so in the future :).

mluessi commented:
How does one build the legacy updater for CC3D? I was trying to test if but can't figure out how to build it.

mlyle commented:
It's part of make all ; `make up_cc3d`

@dustin --- do you think you could do the magic that you did to bring the old Tau stuff in as a ref to bring in the old Brain things?

mluessi commented:
Nice, success on CC3D/Linux upgrading from Renatus

mluessi commented:
@mlyle reported successful tests on Mac, merging.

dustin commented:
Sure, I can add brain refs, too.

On Thu, Mar 24, 2016 at 9:33 AM Martin Luessi notifications@github.com
wrote:

> Merged #801 https://github.com/d-ronin/dRonin/pull/801.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/d-ronin/dRonin/pull/801#event-601997351

dustin commented:
pushed a new tl-mega and brain-mega

On Thu, Mar 24, 2016 at 9:44 AM Dustin Sallings dustin@spy.net wrote:

> Sure, I can add brain refs, too.
> 
> On Thu, Mar 24, 2016 at 9:33 AM Martin Luessi notifications@github.com
> wrote:
> 
> > Merged #801 https://github.com/d-ronin/dRonin/pull/801.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > https://github.com/d-ronin/dRonin/pull/801#event-601997351

mlyle commented:
@dustin -- thx

@mluessi--- did that help?

yamahabest commented:
Win+AQ32 was succesfully tested on another PC, tested a few days ago.

ukmook commented:
Win10 surface 3 tablet Sparky2 tested worked successfully.
